File: cpp/trees/0102-binary-tree-level-order-traversal.cpp
#include <queue>
#include <vector>
#include <structures.h>

using namespace std;


vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> response;
    if (root == nullptr) return response;
        
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        vector<int> level;
        int len = q.size();
        for (int i = 0; i < len; i++) {
            TreeNode *cur = q.front();
            q.pop();
            level.push_back(cur->val);
            if (cur->left) q.push(cur->left);
            if (cur->right) q.push(cur->right);
        }
        response.push_back(level);
    }
    
    return response;
}

File: js/trees/0102-binary-tree-level-order-traversal.js
var levelOrder = function(root) {
    var response = [];
    if (root == undefined) return response;
    
    var queue = [];
    queue.push(root);
    
    while (queue.length > 0) {
        var level = []
        var len = queue.length;
    
        for (var i  = 0; i < len; i++) {
            var cur = queue.shift();
            
            level.push(cur.val);

            if (cur.left) queue.push(cur.left);
            if (cur.right) queue.push(cur.right);
        }
        response.push(level);
    }
    
    return response;
};